anvil/src/eth/backend/db.rs

```diff
@@ -0,0 +1,157 @@
+        let mut info = self.basic(address);
+        info.nonce = nonce;
+        self.insert_account(address, info);
+    }
+
+    /// Sets the balance of the given address
+    fn set_balance(&mut self, address: Address, balance: U256) {
+        let mut info = self.basic(address);
+        info.balance = balance;
+        self.insert_account(address, info);
+    }
+
+    /// Sets the balance of the given address
+    fn set_code(&mut self, address: Address, code: Bytes) {
+        let mut info = self.basic(address);
+        info.code = Some(code.to_vec().into());
+        self.insert_account(address, info);
+    }
+
+    /// Sets the balance of the given address
+    fn set_storage_at(&mut self, address: Address, slot: U256, val: U256);
+}
+
+/// Implement the helper for revm's db
+impl<ExtDB: DatabaseRef + Send + Sync + 'static> Db for CacheDB<ExtDB> {
+    fn insert_account(&mut self, address: Address, account: AccountInfo) {
+        self.insert_cache(address, account)
+    }
+
+    fn set_storage_at(&mut self, address: Address, slot: U256, val: U256) {
+        self.insert_cache_storage(address, slot, val)
+    }
+}
+
+/// Implement the helper for the fork database
+impl Db for ForkedDatabase {
+    fn insert_account(&mut self, address: Address, account: AccountInfo) {
+        self.db.do_insert_account(address, account)
+    }
+
+    fn set_storage_at(&mut self, address: Address, slot: U256, val: U256) {
+        let mut db = self.db.storage.write();
+        db.entry(address).or_default().insert(slot, val);
+    }
+}
+
+/// a [revm::Database] that's forked off another client
+///
+/// The `backend` is used to retrieve (missing) data, which is then fetched from the remote
+/// endpoint. The inner in-memory database holds this storage and will be used for write operations.
+/// This database uses the `backend` for read and the `db` for write operations. But note the
+/// `backend` will also write (missing) data to the `db` in the background
+#[derive(Debug, Clone)]
+pub struct ForkedDatabase {
+    /// responsible for fetching missing data
+    ///
+    /// This is responsible for getting data
+    backend: SharedBackend,
+    /// Contains all the data already fetched
+    ///
+    /// This is used for change commits
+    db: Arc<MemDb>,
+}
+
+impl ForkedDatabase {
+    /// Creates a new instance of this DB
+    pub fn new(backend: SharedBackend, db: Arc<MemDb>) -> Self {
+        Self { backend, db }
+    }
+
+    /// Reset the fork to a fresh forked state, and optionally update the fork config
+    pub fn reset(
+        &self,
+        _url: Option<String>,
+        block_number: Option<u64>,
+    ) -> Result<(), BlockchainError> {
+        if let Some(block_number) = block_number {
+            self.backend
+                .set_pinned_block(block_number)
+                .map_err(|err| BlockchainError::Internal(err.to_string()))?;
+        }
+
+        // TODO need to find a way to update generic provider via url
```

Comment:
the backend handler has a generic `provider: M: Middleware` so there's no easy way to replace it with a new URL. because we don't want to keep the generic type in the sender. But I guess we could fix this with a helper trait that creates a new instance of a provider given a url.